src/Expander.hs

```diff
@@ -115,12 +115,13 @@ expandModule thisMod src = do
   local (set (expanderLocal . expanderModuleName) thisMod) do
     lang <- mustBeModName (view moduleLanguage src)
     initializeLanguage lang
-    modBodyDest <- liftIO $ newModBodyPtr
-    modifyState $ set expanderModuleTop $ Just modBodyDest
+    declTreeDest <- liftIO $ newDeclTreePtr
+    vp <- newDeclValidityPtr
+    modifyState $ set expanderModuleTop $ Just declTreeDest
     decls <- addModuleScope (view moduleContents src)
-    forkExpandDecls modBodyDest decls
+    expandDeclForms declTreeDest mempty vp decls
     expandTasks
```

Comment:
Whenever an `import` is expanded, the tests fail here, at `expandTasks`, with `No progress was possible`. The error message lists a single stuck task: the `ExpandDeclForms` which is waiting for the `import` to complete. I do expect the task list to still contain that task while the `import` is expanded, so I don't understand why `expandTasks` complains. Is that function trying to expand _all_ the tasks? At least one task? As many as possible?

Reply:
That sounds like the import is not filling its validity pointer with something.

That function should be running as many tasks as possible.

Reply:
> That sounds like the import is not filling its validity pointer with something.

`import` doesn't get the chance to fill its validity pointer, because `expandTasks` fails [here](https://github.com/gelisam/klister/pull/37/files#diff-a63cde2fb12e931100bbb12fc30e04eeR582), while we're expanding `import`.

> That function should be running as many tasks as possible.

Then should [this](https://github.com/gelisam/klister/blob/master/src/Expander.hs#L1118) be `return ()` instead of `throwError`? It looks to me like `expandTask` is trying to run as many tasks as possible until there are none left, and if it ever reaches a point where all the remaining tasks are stuck, it fails.

Reply:
changing it to `return ()` allows the tests which use `import` to pass, but now `wait for a signal which is never coming` and `send and receive different signals` are failing, because they are expecting a `NoProgress` failure which is no longer thrown.

Reply:
> It looks to me like expandTask is trying to run as many tasks as possible until there are none left, and if it ever reaches a point where all the remaining tasks are stuck, it fails.

This is exactly what it should do.

Your link labeled "here" doesn't work for me - it just takes me to the overall diff for this PR.

What might be going wrong is that you might not be creating a fresh task ID when necessary. Tasks in the queue don't just get to be structurally equal - they need a notion of identity. This is so that the check for progress can tell the difference between a task that couldn't make progress and a task that made progress (updating the global state somewhere) but left behind something that looks identical to itself.

The `return ()` is definitely not right - we want to detect deadlock, not just skip things arbitrarily in case of deadlock!

Reply:
> This is exactly what it should do.

Hmm, I'm confused. If it should run as many tasks as possible, then I would expect it to run all the tasks it can, continuing if new tasks are forked, and stopping if all the tasks are stuck. Failing if all the available tasks are stuck only seems to make sense if the goal of `expandTasks` is to complete all the tasks which have ever been forked, but I don't understand why we would want that.

> Your link labeled "here" doesn't work for me - it just takes me to the overall diff for this PR.

GitHub's UX is quite poor here; it "helpfully" hides large diffs such as `src/Expander.hs`'s, and so you have to click `load diff` on that file, and then you have to scroll up because the selected line is hidden by the top bar. [Here is another link](https://github.com/gelisam/klister/blob/75362ae938116f06a9ae21c065ba223f736834f4/src/Expander.hs#L576) which doesn't suffer from those problems.

> What might be going wrong is that you might not be creating a fresh task ID when necessary.

I am not creating a fresh task ID, I am using `stillStuck tid task` to re-enqueue the same task when it is stuck, same as the `ExpandMoreDecls` task was doing.

> The `return ()` is definitely not right - we want to detect deadlock, not just skip things arbitrarily in case of deadlock!

I don't think we're skipping _arbitrary_ tasks: we are skipping the tasks which cannot currently make progress, because they are blocked on something else. If this was the top level call into the task system, I agree that we should make sure that all the tasks are expanded and that no leftover tasks remain, as that would indicate that the tasks got permanently stuck and that some of the file is not fully type-checked nor fully expanded. In this case, however, we are importing another file. It might make sense to make sure that all the tasks spawned while expanding the imported file are complete, but I don't think we are equipped to distinguish those tasks from the tasks which were spawned in the file which calls `import`. And I don't think it makes sense to require all the tasks in both files to complete before the `import` task can complete, as there are some already-spawned tasks which are blocked waiting for that `import` to complete, and that's what's causing the failures.

Reply:
I managed to get the tests to pass by distinguishing the places where I do expect all tasks to be completed (after every repl command, when running a file from the command-line, and after each test) from the places where I expect some of the tasks to be left over (when importing a file or initializing the #lang). The former now call `fullyExpandExpr` and `fullyVisit` (there is also an unused `fullyExpandModule`) while the latter call `expandModule` and `visit` (there is also an `expandExpr` which is only used in `fullyExpandExpr`).

Do you still think this is the wrong approach?

Reply:
Unless I've misunderstood you, I do still think this is the wrong approach, for one principled reason and for a heuristic reason.

The principled reason is that modules should be *modular* - they should make sense on their own. At a minimum, a module system should support things like independent type checking/compilation and a separation of interface and implementation. But allowing tasks to remain unsolved after expanding a module means that a task could be left like "Await signal 5, then go do something". Whether or not that thing is ever done could then depend on whether some other module sent the signal. Similar issues could arise from other forms of dependencies. And the `#lang` is just specifying a module from which the default bindings are drawn. So really, when expanding a module, the task queue should start out containing only the top level task, and it should end by requiring the queue to be empty, achieving modularity. Don't recursive calls to the module expander save their task queue and restore it once the recursive expansion is complete?

The heuristic reason for this is that it seems to have worked before, and I don't know why the changes you're making to the system should introduce this behavior, and that makes me suspicious that something else has gone wrong.

Reply:
> Don't recursive calls to the module expander save their task queue and restore it once the recursive expansion is complete?

Ah, that's the piece I was missing! Yes, if the tasks which I claim should still be stuck after the `import` task completes are removed from the task queue before we call `expandTasks`, then I agree that we should expect the task queue to be empty afterwards, we shouldn't expect it to still contain those tasks we removed.

Reply:
What about an operator like this sketch, then?
```
completely :: Expand a -> Expand a
completely act =
  do oldTasks <- getTasks
     setTasks []
     act
     remaining <- getTasks
     if remaining != [] then throwError ... else pure ()
     setTasks oldTasks
```
Then module expansion, recursive or otherwise, can be ``completely (expandModule ...)`` etc. Or `expandModule` can have the `completely` in its definition, because it kinda doesn't make sense to expand a module incompletely here.

Reply:
That's a much better API, thanks.

Reply:
I got rid of the `fully***` functions and am now using `completely` instead. I did not yet change `expandModule` to always use `completely`, but now that the semantics is to save and restore the task queue rather than to fail if the queue is not empty at the end, I think it makes sense to use `completely` a lot more often.

It does not look like we were previously saving and restoring the task queue, I think that's the root cause of the failures. I don't yet understand how the tests were previously passing without this.

Reply:
Okay, so the reason the tests are passing on `master` even though we don't save and restore the task queue is by sheer luck. The order of those two `fork***` calls in `forkExpandDecls` shouldn't matter, and yet the tests fail if I swap them:

```haskell
  forkExpanderTask $
    ExpandMoreDecls restDest sc
      (Syntax (Stx scs loc (List ds)))
      declPhase
  forkExpandOneDecl declDest sc declPhase =<< addRootScope d
```

Ironically, I think those two lines are in that unintuitive order specifically to demonstrate that we don't need to expand the head of the decls before the tail of the decls, we can put those two lines in any order and the task-dependency logic will nevertheless ensure that the tail doesn't start too early. However, since `forkExpanderTask` uses `(task:)` rather than `(++ [task])`, forking the tail task before the head task is precisely the order which leads the expander to run the head task before the tail task.

The next bit of luck is `expandTasks` doesn't pop the tasks from the queue one by one, it pops _all_ the tasks and then runs each of those tasks one at a time. This means that as the very first task gets executed, the global task queue is empty, even though `expandTasks` still has a few other tasks to expand. Since the head `import` task is ran first, the queue starts empty, so its recursive call to `expandTasks` successfully ends with an empty queue as well. When I switch the order, the tail task runs first, it re-enqueues itself, the queue starts full, and so the recursive call to `expandTasks` ends with a full queue and errors out.

The final bit of luck is that a sequence of `import` declarations happens to interleave in such a way that the `import` is always the first task in the queue: first the queue contains the head and the tail decl tasks, we successfully run the head import decl task, and then we successfully run the tail decl task, which pushes the next head and tail decl tasks onto the queue. If there are several `import`s in a row, that next head decl task is another import.

As a conclusion, I have now reached the "How did that ever work?" [stage of debugging](http://plasmasturm.org/log/6debug/) for this particular bug ;)

Reply:
`expandModule` and `initializeLanguage` now both use `completely` (indirectly, via `visit`).

Reply:
Wow! This is some good detective work. I'm really glad you've gotten it cleaned up. Sorry to have built it so strangely to start with.